diff -Bbu -ur ./src/http/ngx_http_upstream.c ../nginx-1.16.0-upstream-jvm-route-pached/src/http/ngx_http_upstream.c
--- ./src/http/ngx_http_upstream.c	2019-08-13 21:51:43.000000000 +0900
+++ ../nginx-1.16.0-upstream-jvm-route-pached/src/http/ngx_http_upstream.c	2019-10-07 19:18:06.033680548 +0900
@@ -5717,6 +5717,8 @@
                                          |NGX_HTTP_UPSTREAM_MAX_CONNS
                                          |NGX_HTTP_UPSTREAM_MAX_FAILS
                                          |NGX_HTTP_UPSTREAM_FAIL_TIMEOUT
+                                         |NGX_HTTP_UPSTREAM_SRUN_ID
+                                         |NGX_HTTP_UPSTREAM_MAX_BUSY
                                          |NGX_HTTP_UPSTREAM_DOWN
                                          |NGX_HTTP_UPSTREAM_BACKUP);
     if (uscf == NULL) {
@@ -5814,9 +5816,9 @@
     ngx_http_upstream_srv_conf_t  *uscf = conf;
 
     time_t                       fail_timeout;
-    ngx_str_t                   *value, s;
+    ngx_str_t                   *value, s, id;
     ngx_url_t                    u;
-    ngx_int_t                    weight, max_conns, max_fails;
+    ngx_int_t                    weight, max_conns, max_fails, max_busy;
     ngx_uint_t                   i;
     ngx_http_upstream_server_t  *us;
 
@@ -5832,7 +5834,10 @@
     weight = 1;
     max_conns = 0;
     max_fails = 1;
+    max_busy = 0;
     fail_timeout = 10;
+    id.data = (u_char *) "a";
+    id.len = sizeof("a") - 1;
 
     for (i = 2; i < cf->args->nelts; i++) {
 
@@ -5881,6 +5886,21 @@
             continue;
         }
 
+        if (ngx_strncmp(value[i].data, "max_busy=", 9) == 0) {
+
+            if (!(uscf->flags & NGX_HTTP_UPSTREAM_MAX_BUSY)) {
+                goto invalid;
+            }
+
+            max_busy = ngx_atoi(&value[i].data[9], value[i].len - 9);
+
+            if (max_busy == NGX_ERROR) {
+                goto invalid;
+            }
+
+            continue;
+        }
+
         if (ngx_strncmp(value[i].data, "fail_timeout=", 13) == 0) {
 
             if (!(uscf->flags & NGX_HTTP_UPSTREAM_FAIL_TIMEOUT)) {
@@ -5899,6 +5919,23 @@
             continue;
         }
 
+        if (ngx_strncmp(value[i].data, "srun_id=", 8) == 0) {
+
+            if (!(uscf->flags & NGX_HTTP_UPSTREAM_SRUN_ID)) {
+                goto invalid;
+            }
+
+            id.len = value[i].len - 8;
+            id.data = &value[i].data[8];
+
+            if (id.len == 0) {
+                goto invalid;
+            }
+
+            continue;
+        }
+
+
         if (ngx_strcmp(value[i].data, "backup") == 0) {
 
             if (!(uscf->flags & NGX_HTTP_UPSTREAM_BACKUP)) {
@@ -5944,7 +5981,9 @@
     us->weight = weight;
     us->max_conns = max_conns;
     us->max_fails = max_fails;
+    us->max_busy = max_busy;
     us->fail_timeout = fail_timeout;
+    us->srun_id = id;
 
     return NGX_CONF_OK;
 
diff -Bbu -ur ./src/http/ngx_http_upstream.h ../nginx-1.16.0-upstream-jvm-route-pached/src/http/ngx_http_upstream.h
--- ./src/http/ngx_http_upstream.h	2019-08-13 21:51:43.000000000 +0900
+++ ../nginx-1.16.0-upstream-jvm-route-pached/src/http/ngx_http_upstream.h	2019-10-07 19:37:01.427112100 +0900
@@ -102,6 +102,9 @@
     ngx_msec_t                       slow_start;
     ngx_uint_t                       down;
 
+    ngx_uint_t                       max_busy;
+    ngx_str_t                        srun_id;
+
     unsigned                         backup:1;
 
     NGX_COMPAT_BEGIN(6)
@@ -116,6 +119,8 @@
 #define NGX_HTTP_UPSTREAM_DOWN          0x0010
 #define NGX_HTTP_UPSTREAM_BACKUP        0x0020
 #define NGX_HTTP_UPSTREAM_MAX_CONNS     0x0100
+#define NGX_HTTP_UPSTREAM_SRUN_ID       0x0040
+#define NGX_HTTP_UPSTREAM_MAX_BUSY      0x0080
 
 
 struct ngx_http_upstream_srv_conf_s {
